frame/transaction-payment/src/lib.rs

```diff
@@ -291,20 +321,26 @@ impl<T: Trait + Send + Sync> SignedExtension for ChargeTransactionPayment<T> whe
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize
 	) -> Result<Self::Pre, TransactionValidityError> {
-		let (_, imbalance) = self.withdraw_fee(who, info, len)?;
-		Ok((self.0, who.clone(), imbalance))
+		let (fee, imbalance) = self.withdraw_fee(who, info, len)?;
+		Ok((self.0, who.clone(), imbalance, fee))
 	}
 
 	fn post_dispatch(
 		pre: Self::Pre,
 		info: &DispatchInfoOf<Self::Call>,
 		post_info: &PostDispatchInfoOf<Self::Call>,
-		_len: usize,
+		len: usize,
 		_result: &DispatchResult,
 	) -> Result<(), TransactionValidityError> {
-		let (tip, who, imbalance) = pre;
+		let (tip, who, imbalance, fee) = pre;
 		if let Some(payed) = imbalance {
-			let refund = Module::<T>::weight_to_fee_with_adjustment(post_info.calc_unspent(info));
+			let actual_fee = Module::<T>::compute_actual_fee(
+				len as u32,
+				info,
+				post_info,
+				tip,
+			);
+			let refund = fee.saturating_sub(actual_fee);
```

Comment:
Isn't it still possible that `fee - actual_fee` is different than `to_fee(weight - actual_weight)`?

Reply:
The `saturating_sub` you marked shouldn't be necessary as `fee` is always larger than `acutal_fee`. Just used it to make it obvious to auditors that no overflow can happen.

> Isn't it still possible that `fee - actual_fee` is different than `to_fee(weight - actual_weight)`?

Yes that is possible. The point of this PR therefore is to not calculate the latter term.

Reply:
But the offchain calculation will do `to_fee(weight - actual_weight)` or not?

Reply:
The off chain calculation does calculate the fee not the refund. So it just does `to_fee(actual_weight)`.

Reply:
Had a discussion with @bkchr  and we came to the conclusion that the off-chain result can still differ:
```
on_chain_fee = to_fee(max_weight) - (to_fee(max_weight) - to_fee(actual_weight)
off_chain_fee = to_fee(actual_weight)
```

The two values can still differ. The PR is not worsening things just shifting the problem off on-chain vs off-chain disparity. Currently looking into ways on how to do only withdraw post dispatch (with previous looking).

Reply:
I think convinced that this is correct. This is how the calculation goes with this PR:
```
on_chain_fee = to_fee(max_weight) - (to_fee(max_weight) - to_fee(actual_weight))
off_chain_fee = to_fee(actual_weight)
```

The `to_fee(max_weight)` cancel each other out cleanly. You are allowed to make this assumption because in the fee world there is only integer arithmetic and no fixed point calculation involved where rounding whould be an issue. I think `on_chain_fee == off_chain_fee` as long as `to_fee(max_weight) > to_fee(actual_weight)` which is true here.

I can be totally wrong but I am unable to come up with counterexample.